csrc/rsa_cipher.cpp

```diff
@@ -76,18 +76,43 @@ JNIEXPORT jint JNICALL Java_com_amazon_corretto_crypto_provider_RsaCipher_cipher
 
         RSA_auto backing; // Used for auto-cleanup
         RSA* r = (RSA *) backing;
+        BIGNUM *bn_n;
+        BIGNUM *bn_e;
+        BIGNUM *bn_d;
+        BIGNUM *bn_p;
+        BIGNUM *bn_q;
+        BIGNUM *bn_dmp1;
+        BIGNUM *bn_dmq1;
+        BIGNUM *bn_iqmp;
         switch (handleMode) {
         case com_amazon_corretto_crypto_provider_RsaCipher_HANDLE_USAGE_IGNORE: // fallthrough
         case com_amazon_corretto_crypto_provider_RsaCipher_HANDLE_USAGE_CREATE:
-            r->n = opt_jarr2bn(env, n);
-            r->e = opt_jarr2bn(env, e);
-            r->d = opt_jarr2bn(env, d);
-            r->p = opt_jarr2bn(env, p);
-            r->q = opt_jarr2bn(env, q);
-            r->dmp1 = opt_jarr2bn(env, dmp1);
-            r->dmq1 = opt_jarr2bn(env, dmq1);
-            r->iqmp = opt_jarr2bn(env, iqmp);
+            bn_n = opt_jarr2bn(env, n);
+            bn_e = opt_jarr2bn(env, e);
+            bn_d = opt_jarr2bn(env, d);
+            bn_p = opt_jarr2bn(env, p);
+            bn_q = opt_jarr2bn(env, q);
+            bn_dmp1 = opt_jarr2bn(env, dmp1);
+            bn_dmq1 = opt_jarr2bn(env, dmq1);
+            bn_iqmp = opt_jarr2bn(env, iqmp);
+
+            if (!bn_e) {
+                bn_e = BN_new();
+            }
+            if (!bn_d) {
+                bn_d = BN_new();
+            }
+            if (!RSA_set0_key(r, bn_n, bn_e, bn_d)) {
+                throw_openssl(EX_RUNTIME_CRYPTO, "Unable to set key parameters");
+            }
 
+            if (p && q && !RSA_set0_factors(r, bn_p, bn_q)) {
```

Comment:
Won't we leak these bignums if we are unable to set them for any reason?

Reply:
You are correct. I'll need to think through the proper clean-up strategy. `_set0_` methods take ownership of the parameters IIRC, so the exact method of handling this needs to be figured out. There are a bunch of hard to test cases here.

Reply:
Probably best to use a `take()` pattern for these `set0`s.

Reply:
I don't think that a `take()` pattern works. I thought about it but the problem is that we don't know if there is a problem until after the method returns. However, we need to `take()` when the parameters are passed in. This means that if the method fails, we've already given up ownership of the objects and thus will leak the memory.

Reply:
The manpage for RSA_set0_factors states:

> Calling this function transfers the memory management of the values to the RSA object, and therefore the values that have been passed in should not be freed by the caller after this function has been called.

This, to me, implies that it owns the objects even on failure. Is this not the case?

Reply:
https://github.com/openssl/openssl/blob/e7fb44e7c3f7a37ff83a6b69ba51a738e549bf5c/crypto/rsa/rsa_lib.c#L334

This particular function can only fail if one of the inputs is NULL, but this results in failure to take ownership. I'll file a bug regarding this unclear documentation.

Reply:
https://github.com/openssl/openssl/issues/10178

Reply:
Thanks for the bug to OpenSSL. Does this mean that the code (as written in ACCP) is correct for memory management?

Reply:
No. If we throw at line 106 we will leak `p`, `q`, `dmp1`, dmq1`, and `iqmp`

Reply:
Please make sure you are looking at the newest version of the code. I've moved entirely over to `BigNumObj` and think that I'm properly managing the memory.

Reply:
Ah, I might be looking at an old diff.